app/services/authentication/authenticator.rb

```diff
@@ -167,5 +173,21 @@ def extract_created_at_from_payload(logged_in_identity)
     def flag_spam_user(user)
       Slack::Messengers::PotentialSpammer.call(user: user)
     end
+
+    def build_identity_from_access_token!
+      # Anyone could figure out a user's 'uid' and impersonate a request to
+      # authenticate in their behalf. The access_token is what we use to
+      # validate the owner of the account actually made the request
+      response = HTTParty.get("https://graph.facebook.com/me?access_token=#{@access_token}")
+      expected_uid = provider.payload.extra.raw_info.id
+      result_uid = JSON.parse(response.body)["id"].to_s
+
+      validation_successful = response.code == 200 && expected_uid == result_uid
+      raise "Authentication error: Access Token validation failed" unless validation_successful
+
+      # Only after validating that the access_token is valid and belongs to the
+      # 'uid' passed in the request is when we can actually build the identity
+      Identity.build_from_access_token(provider)
+    end
```

Comment:
I tried to share the reasoning behind this logic because I thought it was important to make it clear why this is necessary.

We definitely shouldn't trust any request that claims to have authenticated with Facebook for a user id `X`. The "access token authentication requests" send in the `uid` and the `access_token` in the params. This way we can use the `access_token` to make an API call to Facebook to `/me`.

If the user id from FB API response matches with the `uid` in the request we can trust this user has authenticated in the Frontend.

I'm open to feedback here about adding/rephrasing here to make it clearer if necessary 🙂